### PR TITLE
Admin site

### LINKED ISSUE #27: Admin-Seite entwerfen
-> Studenten und Dozenten in Kurse einteilen
-> Studenten in Studentengruppen einteilen
-> Noten von Studenten freigeben

### LINKED ISSUE #49: Seitendesign vereinheitlichen

### LINKED ISSUE #50: Kalender für Kurszeiten

### PR DESCRIPTION
**Pull request by LucRome**
closes #27
closes #49
closes #50

### CODE REVIEW

File: eCourse_backend/courses/urls.py [hunk 1]
```diff
@@ -19,9 +19,22 @@
...
 urlpatterns = [
-    path('overview/', course_overview, name='overview'),
+
     path('create/', create_course, name='create_course'),
     path('alter/<id>', alter_course, name='alter_course'),
     path('delete/<id>', delete_course, name='delte_course'),
-    path('detail/<id>', view_course, name='view_course'),
+    # course
+    path(
+        'course/overview/page<int:page>',
```
Comment (codingHahn) [inline eCourse_backend/courses/urls.py]: Please remove the leading `course`.

File: eCourse_backend/courses/urls.py [hunk 2]
```diff
@@ -19,9 +19,22 @@
...
+        course_overview,
+        name='course_overview'),
+    # create user iframes
+    path('course/iframes/studentlist/<id>', course_student_list_iframe,
```
Comment (codingHahn) [inline eCourse_backend/courses/urls.py]: Please remove the leading `course`.

File: eCourse_backend/courses/urls.py [hunk 3]
```diff
@@ -19,9 +19,22 @@
...
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
+
+    path('edit/<id>', edit_course, name='edit_course'),
+
+    path('course/view/<id>', view_course, name='view_course')
```
Comment (codingHahn) [inline eCourse_backend/courses/urls.py]: Please remove the leading `course`.

File: eCourse_backend/courses/urls.py [hunk 4]
```diff
@@ -19,9 +19,22 @@
...
+        'course/overview/page<int:page>',
+        course_overview,
+        name='course_overview'),
+    # create user iframes
+    path('course/iframes/studentlist/<id>', course_student_list_iframe,
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
+
+    path('edit/<id>', edit_course, name='edit_course'),
```
Comment (codingHahn) [inline eCourse_backend/courses/urls.py:31]: `edit` already exists in `alter`

File: eCourse_backend/courses/urls.py [hunk 5]
```diff
@@ -19,9 +19,22 @@
...
+    # course
+    path(
+        'course/overview/page<int:page>',
+        course_overview,
+        name='course_overview'),
+    # create user iframes
+    path('course/iframes/studentlist/<id>', course_student_list_iframe,
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
```
Comment (codingHahn) [inline eCourse_backend/courses/urls.py]: Why do we need an additional admin course creation?

File: eCourse_backend/courses/views.py [hunk 6]
```diff
@@ -1,25 +1,82 @@
...
+@login_required
+def course_overview(request, page=1):
+
     user_id = request.user.id
     if request.method == 'GET':
         # officer has type 3 in db
         if request.user.type == 3:
             courses = Course.objects.all()
         else:
+            # courses = Course.objects.filter(student = user_id)
```
Comment (codingHahn) [inline eCourse_backend/courses/views.py]: Please remove the commented out code

File: eCourse_backend/courses/views.py [hunk 7]
```diff
@@ -1,25 +1,82 @@
...
+    }
+    return render(request, 'courses/overview.html', context)
+
+
+@login_required
+def view_course(request, id):
+    if request.method == 'GET':
+        course = get_object_or_404(Course, pk=id)
+        lecturer = course.lecturer_id
+        # print('lecturer', lecturer)
```
Comment (codingHahn) [inline eCourse_backend/courses/views.py]: Please remove the commented out code

File: eCourse_backend/courses/views.py [hunk 8]
```diff
@@ -60,6 +117,26 @@ def create_course(request):
...
+    else:
+        form = CourseForm()
+
+    context = {
+        'form': form,
+        'success': success
+    }
+    return render(request, 'courses/create_course_admin.html', context)
+
+
```
Comment (codingHahn) [inline eCourse_backend/courses/views.py]: This seems very similar to `create_course`. What functionality do you miss from the existing function?

File: eCourse_backend/courses/views.py [hunk 9]
```diff
@@ -82,3 +159,17 @@ def delete_course(request, id):
...
+def edit_course(request, id):
+    if request.method == 'POST':
+        form = CourseForm(request.POST)
+        if form.is_valid():
+            form.save()
+    else:
+        course_object = get_object_or_404(Course, pk=id)
+        form = CourseForm(model_to_dict(course_object))
+
+    return render(request, 'admin/edit_course.html', {'form': form})
```
Comment (codingHahn) [inline eCourse_backend/courses/views.py]: `alter_course` already does the same thing

File: eCourse_backend/eCourse_backend/settings.py [hunk 10]
```diff
@@ -43,11 +43,11 @@
     'django.contrib.messages',
     'django.contrib.staticfiles',
     'django_extensions',
-    'widget_tweaks',
     'eCourse_backend',
     'users',
     'file_exchange',
     'courses',
+    'widget_tweaks'
```
Comment (codingHahn) [inline eCourse_backend/eCourse_backend/settings.py:50]: Is this necessary? If not, please revert this change. It pollutes the history

File: eCourse_backend/eCourse_backend/static/Css/basic_definitions.css [hunk 11]
```diff
@@ -0,0 +1,31 @@
+.input {
```
Comment (codingHahn) [inline eCourse_backend/eCourse_backend/static/Css/basic_definitions.css:1]: Please use lower case names for directories

File: eCourse_backend/eCourse_backend/static/readme.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+* **/Css:**
+    * Folder contains css files that are served as static files
+* **./:**
+    * Contains:
+        * different Versions of the Logo + some other files
+            * used to create compability with different browser features (some are just for Android/Windows/...)
+            * other files are also needed for these features (e.g. site.webmanifest)
+            * created with: [RealFaviconGenerator](https://realfavicongenerator.net/)
+        * the original logo
+            * created with: [FreeLogoDesign](https://de.freelogodesign.org)
```
Comment (codingHahn) [inline eCourse_backend/eCourse_backend/static/readme.md:12]: This file is publically accessible. Is that wanted?

File: eCourse_backend/templates/admin/edit_course.html [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+<form method="post">
+  {% csrf_token %}
+  {% comment %} Button: Zurück zur Übersicht {% endcomment %}
+  <a href="{% url 'home' %}"><button class="btn btn-danger" type="button">Zurück</button></a>
+  {% comment %} Textbox: Kurs Daten {% endcomment %}
+  {% include 'forms/bs4_text_form.html' with form=form %}
+  {% comment %} Button: Änderungen Speichern {% endcomment %}
+  <button class="btn btn-danger" type="submit">Speichern</button>
+</form>
+{% endblock %}
```
Comment (codingHahn) [inline eCourse_backend/templates/admin/edit_course.html]: We already have `alter_courses`

File: eCourse_backend/templates/admin/home_Admin.html [hunk 14]
```diff
@@ -0,0 +1,5 @@
+{% extends 'home/home_auth.html' %}
```
Comment (codingHahn) [inline eCourse_backend/templates/admin/home_Admin.html:1]: Please use lower case file names

File: eCourse_backend/templates/admin/users/iframes/create_user/create_template.html [hunk 15]
```diff
@@ -0,0 +1,34 @@
+{% extends 'iframe_template.html' %}
```
Comment (codingHahn) [inline eCourse_backend/templates/admin/users/iframes/create_user/create_template.html]: Could you rename this file? To me at least, it isn't clear what `create_template` is.

File: eCourse_backend/templates/base.html [hunk 16]
```diff
@@ -6,11 +6,20 @@
 <html>
 
 <head>
-  <meta charset="utf-8">
+    <meta charset="utf-8">
```
Comment (codingHahn) [inline eCourse_backend/templates/base.html:9]: Please fix indentation

File: eCourse_backend/templates/home/home_notAuth.html [hunk 17]
```diff
@@ -0,0 +1,12 @@
+{% extends 'base.html' %}
```
Comment (codingHahn) [inline eCourse_backend/templates/home/home_notAuth.html:1]: Please use only lower case file names

File: eCourse_backend/users/urls.py [hunk 18]
```diff
@@ -19,8 +19,34 @@
...
+    # studentlist iframe
+    path('admin_s/iframes/studentlist/page<int:page>',
+         student_list_iframe, name='studentlist_admin_iframe'),
+    # lecturerlist iframe
+    path('admin_s/iframes/lecturerlist/page<int:page>',
+         lecturer_list_iframe, name='lecturerlist_admin_iframe'),
+    # admin + staff list (iframe)
+    path('admin_s/iframes/adminstafflist/page<int:page>',
+         staff_admin_list_iframe, name='adminlist_admin_iframe'),
 ]
```
Comment (codingHahn) [inline eCourse_backend/users/urls.py]: Please revise the site structure. We don't need extra `admin_s` functions. The normal `create` etc functions can only be accessed by an office user (admin) anyways. What is the greater goal here? But I understand the value of having extra pages for all user types.

We also need to talk about the `iframe/` url path naming. I think that is a good idea, but if we implement this, we should make this change in the whole codebase at once.

File: eCourse_backend/users/views.py [hunk 19]
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
...
+
+@login_required
+@permission_required('users.manage_users', raise_exception=True)
+def user_administration_admin(request):
+    if request.method == 'GET':
+        return render(request, 'admin/users/administration.html', {})
+
+# IFrames
+# create lecturer
+
```
Comment (codingHahn) [inline eCourse_backend/users/views.py]: Isn't that exactly the same as the `users/overview` page?

File: eCourse_backend/users/views.py [hunk 20]
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
...
+    user_to_delete = get_object_or_404(User, username=username)
+    # TODO: catch if delete operation fails (?)
+    user_to_delete.delete()
+    context = {
+        'username': username,
+    }
+    return render(
+        request,
+        'admin/users/iframes/deleted_user_iframe.html',
+        context)
```
Comment (codingHahn) [inline eCourse_backend/users/views.py]: This functionality already exists in `delete_user`

File: eCourse_backend/users/views.py [hunk 21]
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
...
+    context = {
+        'username': username,
+        'user_form': form,
+        'update_success': update_success
+    }
+
+    return render(
+        request,
+        'admin/users/iframes/edit_user_modalcontent.html',
+        context)
```
Comment (codingHahn) [inline eCourse_backend/users/views.py]: Please state which use cases the already existing view `alter_users` did not fullfill. Please extend the existing function

File: eCourse_backend/users/views.py [hunk 22]
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
...
+    page_obj = paginator.get_page(page)
+
+    context = {
+        'page_obj': page_obj,
+        'filter_form': filter_form,
+    }
+    return render(
+        request,
+        'admin/users/iframes/staff_admin_list.html',
+        context)
```
Comment (codingHahn) [inline eCourse_backend/users/views.py]: The three views are very similar. Could this be merged into one view?

File: eCourse_backend/users/views.py [hunk 23]
```diff
@@ -60,3 +65,256 @@ def alter_user(request, id):
...
+        user_form = StaffForm()
+    context = {
+        'user_form': user_form,
+        'success': save_success,
+    }
+
+    return render(
+        request,
+        'admin/users/iframes/create_user/create_officeuser.html',
+        context)
```
Comment (codingHahn) [inline eCourse_backend/users/views.py]: The three views are very similar. Could this be merged into one view?
Furthermore, the `create_user` function has much of the same functionality in fewer lines of code

File: eCourse_backend/users/forms.py [hunk 24]
```diff
@@ -1,8 +1,36 @@
...
+    matr_nr = IntegerField(label="Matrikel Nummer", required=False)
+    username = CharField(max_length=50, label="Username", required=False)
+    first_name = CharField(max_length=50, label="Vorname", required=False)
+    last_name = CharField(max_length=50, label="Nachname", required=False)
+
+
+class AdminStaffFilterForm(Form):
+    username = CharField(max_length=50, label="Username", required=False)
+    first_name = CharField(max_length=50, label="Vorname", required=False)
+    last_name = CharField(max_length=50, label="Nachname", required=False)
```
Comment (codingHahn) [inline eCourse_backend/users/forms.py]: Those forms are very similar. Could they be merged?

File: eCourse_backend/users/forms.py [hunk 25]
```diff
@@ -1,8 +1,36 @@
...
         model = User
         fields = ['username', 'first_name', 'last_name', 'email', 'matr_nr']
+
+
+class StaffForm(ModelForm):
+    class Meta:
+        model = User
+        fields = ['username', 'first_name', 'last_name',
+                  'email', 'matr_nr', 'is_superuser']
+
```
Comment (codingHahn) [inline eCourse_backend/users/forms.py]: Please don't declare regular users as superusers, because they can put the database in a bad state. They are meant for maintenance tasks.

**Review by codingHahn — CHANGES_REQUESTED** [review]
Please address the points mentioned in the review. Please remember that filenames and directories should be snake case.

Please do smaller PRs in the future.

File: eCourse_backend/eCourse_backend/static/readme.md (hunk 12 above)
Comment (LucRome) [reply eCourse_backend/eCourse_backend/static/readme.md:12]: At the moment it doesn't hurt and otherwise it would be quite confusing why these files are there
I would leave it there since djangos static file hosting shouldn't be used in a production environment anyways

File: eCourse_backend/courses/urls.py (hunk 4 above)
Comment (LucRome) [reply eCourse_backend/courses/urls.py:31]: should be resolved

File: eCourse_backend/users/views.py (hunk 20 above)
Comment (LucRome) [reply eCourse_backend/users/views.py]: should be resolved

File: eCourse_backend/users/forms.py [hunk 26]
```diff
@@ -1,8 +1,27 @@
...
     class Meta:
         model = User
         fields = ['username', 'first_name', 'last_name', 'email', 'matr_nr']
+
+
+class StaffForm(ModelForm):
+    class Meta:
+        model = User
+        fields = ['username', 'first_name', 'last_name',
+                  'email', 'matr_nr']
```
Comment (phie73) [inline eCourse_backend/users/forms.py]: these two forms are exactly the same

File: eCourse_backend/users/urls.py [hunk 27]
```diff
@@ -19,8 +19,29 @@
...
+         name='edituser_admin_modalcontent_iframe'),
+    # studentlist iframe
+    path('admin_s/iframes/studentlist/page<int:page>',
+         student_list_iframe, name='studentlist_admin_iframe'),
+    # lecturerlist iframe
+    path('admin_s/iframes/lecturerlist/page<int:page>',
+         lecturer_list_iframe, name='lecturerlist_admin_iframe'),
+    # admin + staff list (iframe)
+    path('admin_s/iframes/adminstafflist/page<int:page>',
+         staff_admin_list_iframe, name='adminlist_admin_iframe'),
```
Comment (phie73) [inline eCourse_backend/users/urls.py]: please remove admin_s

File: eCourse_backend/users/views.py [hunk 28]
```diff
@@ -1,62 +1,258 @@
...
+from django.views.decorators.clickjacking import xframe_options_exempt
+from django.core.paginator import Paginator
 
 # Create your views here.
 
+# Views given from Backend, use as orientation
+
+
+# Views for the admin
+# TODO: check whethter user is from the right group
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please remove the todo because the permissions already solve this point

File: eCourse_backend/users/views.py [hunk 29]
```diff
@@ -1,62 +1,258 @@
...
 @login_required
-@permission_required('users.manage_users')
-def overview(request):
-    # TODO: Filter Users based on attributes (group, course, etc)
+@permission_required('users.manage_users', raise_exception=True)
+def user_administration_admin(request):
     if request.method == 'GET':
-        all_users = User.objects.all()
-        return render(request, 'users/overview.html', {'users': all_users})
+        return render(request, 'admin/users/administration.html', {})
```
Comment (phie73) [inline eCourse_backend/users/views.py]: by the function name their purpose isn't clear please rename

**Comment by codingHahn** [comment]
When I log in this happens: 
image

File: eCourse_backend/courses/urls.py [hunk 30]
```diff
@@ -19,9 +19,14 @@
...
-    path('alter/<id>', alter_course, name='alter_course'),
     path('delete/<id>', delete_course, name='delte_course'),
-    path('detail/<id>', view_course, name='view_course'),
+    # course
+    path('overview/page<int:page>',course_overview,name='course_overview'),
+    # create user iframes
+    path('iframes/studentlist/<id>', course_student_list_iframe,
+         name='course_studentlist_iframe'),
+    # admin: create course
+    path('admin/create/', create_course_admin, name="create_course_admin"),
```
Comment (phie73) [inline eCourse_backend/courses/urls.py]: please remove the admin from 'admin/creare/'

File: eCourse_backend/courses/views.py [hunk 31]
```diff
@@ -1,16 +1,51 @@
...
+        # print('lecturer', lecturer)
+        students = course.student.all()
+
+    context = {
+        'filter_form': filter_form,
+        'lecturer': lecturer,
+        'students': students,
+    }
+    return render(request, 'courses/iframes/course_student_list.html', context)
+
```
Comment (phie73) [inline eCourse_backend/courses/views.py]: this function already exists on the master branch see courses/views.py view_course on the master branch. The functionality is extended for mapping the files and exercises to the courses. This function used to exist in this file (see below) but it was deleted. If you need additional data please add this to the existing function or create an issue.

File: eCourse_backend/courses/views.py [hunk 32]
```diff
@@ -19,50 +54,50 @@ def course_overview(request):
...
+    context = {
+        'page_obj': page_obj,
+    }
+    return render(request, 'courses/overview.html', context)
 
 
 @login_required
 @permission_required('courses.create_course', raise_exception=True)
-def create_course(request):
+def create_course_admin(request):
```
Comment (phie73) [inline eCourse_backend/courses/views.py]: please revert the name change or explain why it is needed.

File: eCourse_backend/users/views.py [hunk 33]
```diff
@@ -1,62 +1,258 @@
...
+        'admin/users/iframes/edit_user_modalcontent.html',
+        context)
+
+# Delete user Iframe
+
+
+@xframe_options_exempt
+@login_required
+@permission_required('users.manage_users', raise_exception=True)
+def delete_user_iframe(request, username):
```
Comment (phie73) [inline eCourse_backend/users/views.py]: Please keep the user id instead of the username for consistency reasons

File: eCourse_backend/users/views.py [hunk 34]
```diff
@@ -1,62 +1,258 @@
...
+# Edit User IFrame
+
+# TODO: correctly edit user
+
+
+@xframe_options_exempt
 @login_required
 @permission_required('users.manage_users', raise_exception=True)
-def alter_user(request, id):
+def edit_user_admin_modalcontent_iframe(request, username):
```
Comment (phie73) [inline eCourse_backend/users/views.py]: Please keep the user id instead of the username for consistency reasons.

**Review by phie73 — CHANGES_REQUESTED** [review]

File: eCourse_backend/courses/views.py [hunk 35]
```diff
@@ -1,16 +1,51 @@
...
+            students = Student.objects.filter(
+                matr_nr__contains=filter_form['matr_nr'].data,
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
+    elif request.method == 'GET':
+        filter_form = CourseStudentFilterForm()
+        course = get_object_or_404(Course, pk=id)
+        lecturer = course.lecturer
+        # print('lecturer', lecturer)
```
Comment (phie73) [inline eCourse_backend/courses/views.py]: could you please remove this debug info?

File: eCourse_backend/users/views.py [hunk 36]
```diff
@@ -1,62 +1,257 @@
...
+    save_success = False
     if request.method == 'POST':
         user_form = UserForm(request.POST)
         if user_form.is_valid():
-            user_form.save()
+            Lecturer.objects.create_user(
+                username=user_form['username'].data,
+                email=user_form['email'].data,
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

File: eCourse_backend/users/views.py [hunk 37]
```diff
@@ -1,62 +1,257 @@
...
+def create_officeuser_iframe(request):
+    save_success = False
+    if request.method == 'POST':
+        user_form = UserForm(request.POST)
+        if user_form.is_valid():
+            Office.objects.create_user(
+                username=user_form['username'].data,
+                email=user_form['email'].data,
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

File: eCourse_backend/users/views.py [hunk 38]
```diff
@@ -1,62 +1,257 @@
...
+    if request.method == 'POST':
+        user_form = StudentForm(request.POST)
+        if user_form.is_valid():
+            Student.objects.create_user(
+                username=user_form['username'].data,
+                email=user_form['email'].data,
+                first_name=user_form['first_name'].data,
+                last_name=user_form['last_name'].data,
+                matr_nr=user_form['matr_nr'].data
+            )
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

File: eCourse_backend/users/views.py [hunk 39]
```diff
@@ -1,62 +1,257 @@
...
+    if request.method == 'POST':
+        filter_form = StudentFilterForm(
+            request.POST, )
+        if filter_form.is_valid():
+            # Filter
+            students = Student.objects.filter(
+                matr_nr__contains=filter_form['matr_nr'].data,
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

File: eCourse_backend/users/views.py [hunk 40]
```diff
@@ -1,62 +1,257 @@
...
+def lecturer_list_iframe(request, page=1):
+    if request.method == 'POST':
+        filter_form = LecturerAndOfficeFilterForm(
+            request.POST, )
+        if filter_form.is_valid():
+            # Filter
+            lecturers = Lecturer.objects.filter(
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

File: eCourse_backend/users/views.py [hunk 41]
```diff
@@ -1,62 +1,257 @@
...
     if request.method == 'POST':
-        form = UserForm(request.POST)
+        filter_form = LecturerAndOfficeFilterForm(
+            request.POST, )
+        if filter_form.is_valid():
+            # Filter
+            users = Office.objects.filter(
+                first_name__contains=filter_form['first_name'].data,
+                last_name__contains=filter_form['last_name'].data,
+                username__contains=filter_form['username'].data)
```
Comment (phie73) [inline eCourse_backend/users/views.py]: please use validated data, this is vulnerable looking at sql injections

**Review by phie73 — CHANGES_REQUESTED** [review]
looks good overall, please address the minor issues we found. Looking forward to merge the pr when this small changes are done.

**Review by noahhauke22 — APPROVED** [review]

File: eCourse_backend/users/views.py (hunk 36 above)
Comment (LucRome) [reply eCourse_backend/users/views.py]: How could that be achieved??
Could you please demonstrate the changes as I don't know exactly how it should be done

File: eCourse_backend/users/views.py (hunk 41 above)
Comment (LucRome) [reply eCourse_backend/users/views.py]: Also please at least provide an Example, I don't know how else it could be done
And also what do you mean by validated data? Since we call the .is_valid() function.

File: eCourse_backend/users/views.py (hunk 36 above)
Comment (phie73) [reply eCourse_backend/users/views.py]: we posted an example in the discord channel, sorry for not mentioning here

**Review by phie73 — APPROVED** [review]
lets solve the merge conflicts together on Monday?

**Comment by LucRome** [comment]
merge conflicts should be solved:

- the Logo.png should be the newest Version
- in `courses/views.py` there was
    - a conflict regarding the imports -> all imports were kept
    - a conflict regarding the two view function for the course overview
        -> both functions kept (also the one from the backend, since it introduced new functions, which should be used by the frontend, should be merged and deleted later on)
- the template `courses/detail.html` was kept, despite not being used by frontend, since it also introduced some new functions

**Review by codingHahn — APPROVED** [review]
